File: services/plugin-service/src/plugin/plugin.service.ts
import { randomUUID } from 'node:crypto';
import { readFile } from 'node:fs/promises';
import { join } from 'node:path';

import {
  BadRequestException,
  Injectable,
  Logger,
  NotFoundException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { DatabaseService } from '../database/database.service';
import { MetricsService } from '../metrics/metrics.service';
import { RequestContextService } from '../observability/request-context.service';
import {
  ExecutionRow,
  PluginRow,
  PluginStatus,
} from './entities/plugin.entity';
import { RegisterPluginDto } from './dto/register-plugin.dto';
import { QueryPluginDto } from './dto/query-plugin.dto';
import { UpdatePluginDto } from './dto/update-plugin.dto';
import { ExecutePluginDto } from './dto/execute-plugin.dto';
import { SandboxResult, SandboxService } from './sandbox.service';

@Injectable()
export class PluginService {
  private readonly logger = new Logger(PluginService.name);
  private readonly storeDirectory: string;

  constructor(
    private readonly db: DatabaseService,
    private readonly config: ConfigService,
    private readonly sandbox: SandboxService,
    private readonly metrics: MetricsService,
    private readonly requestContext: RequestContextService,
  ) {
    this.storeDirectory = this.config.get<string>('plugin.storeDirectory') ?? 'plugins';
  }

  async register(dto: RegisterPluginDto): Promise<PluginRow> {
    const id = randomUUID();
    const tenantId = this.requestContext.tenantId ?? dto.tenantId;

    const result = await this.db.query<PluginRow>(
      `INSERT INTO plugins (
        id, name, version, description, author, license, min_app_version,
        status, permissions, capabilities, hooks, tags, entry_point,
        tenant_id, homepage, repository, icon, config_schema, code
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, 'REGISTERED', $8::jsonb, $9, $10, $11, $12, $13, $14, $15, $16, $17::jsonb, $18)
      RETURNING *`,
      [
        id,
        dto.name,
        dto.version,
        dto.description,
        dto.author,
        dto.license,
        dto.minAppVersion,
        JSON.stringify(dto.permissions ?? {}),
        dto.capabilities,
        dto.hooks,
        dto.tags,
        dto.entryPoint,
        tenantId,
        dto.homepage ?? null,
        dto.repository ?? null,
        undefined,
        dto.configSchema ? JSON.stringify(dto.configSchema) : null,
        dto.code ?? null,
      ],
    );

    this.metrics.activePlugins.inc();
    this.logger.log({ pluginId: id }, 'Plugin registered');
    return result.rows[0]!;
  }

  async findAll(queryDto: QueryPluginDto): Promise<{ data: PluginRow[]; meta: Record<string, unknown> }> {
    const page = queryDto.page ?? 1;
    const pageSize = queryDto.pageSize ?? 20;
    const offset = (page - 1) * pageSize;
    const conditions: string[] = ['1 = 1'];
    const params: unknown[] = [];
    let paramIndex = 1;

    if (queryDto.status) {
      conditions.push(`p.status = $${paramIndex}`);
      params.push(queryDto.status);
      paramIndex++;
    }

    if (queryDto.tags && queryDto.tags.length > 0) {
      conditions.push(`p.tags && $${paramIndex}::text[]`);
      params.push(queryDto.tags);
      paramIndex++;
    }

    if (queryDto.search) {
      conditions.push(`(p.name ILIKE $${paramIndex} OR p.description ILIKE $${paramIndex})`);
      params.push(`%${queryDto.search}%`);
      paramIndex++;
    }

    const whereClause = conditions.join(' AND ');

    const countResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM plugins p WHERE ${whereClause}`,
      params,
    );
    const total = Number(countResult.rows[0]!.count);

    const dataResult = await this.db.query<PluginRow>(
      `SELECT p.* FROM plugins p WHERE ${whereClause}
       ORDER BY p.created_at DESC
       LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`,
      [...params, pageSize, offset],
    );

    const totalPages = Math.ceil(total / pageSize);
    return {
      data: dataResult.rows,
      meta: { page, pageSize, total, totalPages, hasNext: page < totalPages, hasPrevious: page > 1 },
    };
  }

  async findById(id: string): Promise<PluginRow> {
    const result = await this.db.query<PluginRow>(
      `SELECT * FROM plugins WHERE id = $1`,
      [id],
    );
    if (result.rows.length === 0) {
      throw new NotFoundException(`Plugin with id "${id}" not found`);
    }
    return result.rows[0]!;
  }

  async update(id: string, dto: UpdatePluginDto): Promise<PluginRow> {
    const existing = await this.findById(id);

    const version = dto.version ?? existing.version;
    const description = dto.description ?? existing.description;
    const entryPoint = dto.entryPoint ?? existing.entry_point;
    const permissions = dto.permissions
      ? JSON.stringify(dto.permissions)
      : existing.permissions;
    const capabilities = dto.capabilities ?? existing.capabilities;
    const hooks = dto.hooks ?? existing.hooks;
    const tags = dto.tags ?? existing.tags;
    const status = dto.status ?? existing.status;

    const result = await this.db.query<PluginRow>(
      `UPDATE plugins SET
        version = $1, description = $2, entry_point = $3, permissions = $4::jsonb,
        capabilities = $5, hooks = $6, tags = $7, status = $8, updated_at = NOW()
       WHERE id = $9
       RETURNING *`,
      [version, description, entryPoint, permissions, capabilities, hooks, tags, status, id],
    );

    this.logger.log({ pluginId: id, version }, 'Plugin updated');
    return result.rows[0]!;
  }

  async remove(id: string): Promise<{ message: string }> {
    const result = await this.db.query(
      `DELETE FROM plugins WHERE id = $1`,
      [id],
    );
    if (result.rowCount === 0) {
      throw new NotFoundException(`Plugin with id "${id}" not found`);
    }
    this.metrics.activePlugins.dec();
    this.logger.log({ pluginId: id }, 'Plugin unregistered');
    return { message: 'Plugin unregistered successfully' };
  }

  async getPermissions(id: string): Promise<{ permissions: Record<string, unknown> }> {
    const plugin = await this.findById(id);
    return { permissions: JSON.parse(plugin.permissions) };
  }

  async getExecutions(
    id: string,
    page = 1,
    pageSize = 20,
  ): Promise<{ data: ExecutionRow[]; meta: Record<string, unknown> }> {
    await this.findById(id);
    const offset = (page - 1) * pageSize;

    const countResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM plugin_executions WHERE plugin_id = $1`,
      [id],
    );
    const total = Number(countResult.rows[0]!.count);

    const dataResult = await this.db.query<ExecutionRow>(
      `SELECT * FROM plugin_executions WHERE plugin_id = $1
       ORDER BY created_at DESC LIMIT $2 OFFSET $3`,
      [id, pageSize, offset],
    );

    const totalPages = Math.ceil(total / pageSize);
    return {
      data: dataResult.rows,
      meta: { page, pageSize, total, totalPages },
    };
  }

  async execute(id: string, dto: ExecutePluginDto): Promise<ExecutionRow> {
    const plugin = await this.findById(id);

    if (plugin.status === PluginStatus.DISABLED) {
      throw new BadRequestException(`Plugin "${id}" is disabled`);
    }

    const code = dto.code ?? plugin.code ?? (await this.loadCodeFromDisk(plugin.entry_point));
    if (!code) {
      throw new BadRequestException(
        `Plugin "${id}" has no executable code (set code or entryPoint referencing a file in the store)`,
      );
    }

    const timeoutMs = dto.timeoutMs ?? this.config.get<number>('plugin.sandboxTimeoutMs') ?? 5000;

    const result: SandboxResult = await this.sandbox.run(
      code,
      {
        input: dto.input ?? {},
        tenantId: plugin.tenant_id,
        pluginId: plugin.id,
        permissions: JSON.parse(plugin.permissions),
      },
      timeoutMs,
    );

    const execId = randomUUID();
    const insertResult = await this.db.query<ExecutionRow>(
      `INSERT INTO plugin_executions (id, plugin_id, status, duration_ms, output, error_message)
       VALUES ($1, $2, $3, $4, $5::jsonb, $6)
       RETURNING *`,
      [
        execId,
        plugin.id,
        result.status,
        result.durationMs,
        JSON.stringify(result.output ?? null),
        result.errorMessage ?? null,
      ],
    );

    this.metrics.sandboxExecutions.inc({ plugin_id: plugin.id, status: result.status });
    if (result.status !== 'success') {
      this.metrics.sandboxFailures.inc({
        plugin_id: plugin.id,
        reason: result.status,
      });
    }

    if (result.status === 'success') {
      await this.db.query(
        `UPDATE plugins SET status = 'ACTIVE', updated_at = NOW() WHERE id = $1`,
        [plugin.id],
      );
    }

    this.logger.log(
      { pluginId: id, status: result.status, durationMs: result.durationMs },
      'Plugin executed',
    );

    return insertResult.rows[0]!;
  }

  private async loadCodeFromDisk(entryPoint: string): Promise<string | null> {
    if (!entryPoint) return null;
    try {
      const filePath = join(process.cwd(), this.storeDirectory, entryPoint);
      return await readFile(filePath, 'utf-8');
    } catch (err) {
      this.logger.warn(
        { entryPoint, err: (err as Error).message },
        'Failed to read plugin code from disk',
      );
      return null;
    }
  }
}

File: apps/backend/src/backend/services/plugin_service.py
import json
import time
import uuid
from datetime import datetime, timezone

from fastapi import HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession


class PluginService:
    async def register(self, dto, tenant_id: str | None, db: AsyncSession):
        plugin_id = uuid.uuid4()
        now = datetime.now(timezone.utc)
        permissions_json = json.dumps(dto.permissions.model_dump() if hasattr(dto.permissions, "model_dump") else dto.permissions)
        tags_list = f"{{{','.join(f'\"{t}\"' for t in dto.tags)}}}" if dto.tags else "{}"
        capabilities_list = f"{{{','.join(f'\"{c}\"' for c in dto.capabilities)}}}" if dto.capabilities else "{}"
        hooks_list = f"{{{','.join(f'\"{h}\"' for h in dto.hooks)}}}" if dto.hooks else "{}"

        stmt = text("""
            INSERT INTO plugins (id, name, version, author, description, license, status, permissions, capabilities, hooks, tags, entry_point, tenant_id, homepage, repository, icon, config_schema, code, min_app_version, created_at, updated_at)
            VALUES (:id, :name, :version, :author, :description, :license, 'REGISTERED', :permissions::jsonb, :capabilities::text[], :hooks::text[], :tags::text[], :entry_point, :tenant_id, :homepage, :repository, :icon, :config_schema::jsonb, :code, :min_app_version, :created_at, :updated_at)
            RETURNING id, name, version, author, description, license, status, permissions, capabilities, hooks, tags, entry_point, tenant_id, homepage, repository, icon, config_schema, code, min_app_version, created_at, updated_at
        """)
        result = await db.execute(stmt, {
            "id": plugin_id,
            "name": dto.name,
            "version": dto.version,
            "author": dto.author,
            "description": dto.description,
            "license": dto.license,
            "min_app_version": dto.min_app_version,
            "permissions": permissions_json,
            "capabilities": dto.capabilities or [],
            "hooks": dto.hooks or [],
            "tags": dto.tags,
            "entry_point": dto.entry_point,
            "tenant_id": tenant_id,
            "homepage": dto.homepage,
            "repository": dto.repository,
            "icon": dto.icon,
            "config_schema": json.dumps(dto.config_schema) if dto.config_schema else None,
            "code": dto.code,
            "created_at": now,
            "updated_at": now,
        })
        row = result.mappings().first()
        return dict(row) if row else None

    async def list_plugins(self, page: int, page_size: int, status: str | None, tags: list[str] | None, search: str | None, tenant_id: str | None, db: AsyncSession):
        conditions = ["1=1"]
        params = {}

        if tenant_id:
            conditions.append("p.tenant_id = :tenant_id")
            params["tenant_id"] = tenant_id
        if status:
            conditions.append("p.status = :status")
            params["status"] = status
        if tags:
            conditions.append("p.tags && :tags")
            params["tags"] = tags
        if search:
            conditions.append("(p.name ILIKE :search OR p.description ILIKE :search OR p.author ILIKE :search)")
            params["search"] = f"%{search}%"

        where_clause = " AND ".join(conditions)

        count_sql = text(f"SELECT COUNT(*) FROM plugins p WHERE {where_clause}")
        count_result = await db.execute(count_sql, params)
        total = count_result.scalar_one()

        offset = (page - 1) * page_size
        data_sql = text(f"""
            SELECT p.id, p.name, p.version, p.author, p.description, p.license, p.status,
                   p.permissions, p.capabilities, p.hooks, p.tags, p.entry_point,
                   p.tenant_id, p.homepage, p.repository, p.icon, p.config_schema, p.code,
                   p.min_app_version, p.created_at, p.updated_at
            FROM plugins p
            WHERE {where_clause}
            ORDER BY p.created_at DESC
            LIMIT :limit OFFSET :offset
        """)
        params["limit"] = page_size
        params["offset"] = offset
        data_result = await db.execute(data_sql, params)
        rows = data_result.mappings().all()
        return [dict(r) for r in rows], total

    async def get_plugin(self, plugin_id: uuid.UUID, db: AsyncSession):
        stmt = text("""
            SELECT id, name, version, author, description, license, status,
                   permissions, capabilities, hooks, tags, entry_point,
                   tenant_id, homepage, repository, icon, config_schema, code,
                   min_app_version, created_at, updated_at
            FROM plugins WHERE id = :id
        """)
        result = await db.execute(stmt, {"id": plugin_id})
        row = result.mappings().first()
        return dict(row) if row else None

    async def update_plugin(self, plugin_id: uuid.UUID, dto, db: AsyncSession):
        sets = []
        params = {"id": plugin_id}

        if dto.version is not None:
            sets.append("version = :version")
            params["version"] = dto.version
        if dto.description is not None:
            sets.append("description = :description")
            params["description"] = dto.description
        if dto.entry_point is not None:
            sets.append("entry_point = :entry_point")
            params["entry_point"] = dto.entry_point
        if dto.permissions is not None:
            sets.append("permissions = :permissions::jsonb")
            perms = dto.permissions.model_dump() if hasattr(dto.permissions, "model_dump") else dto.permissions
            params["permissions"] = json.dumps(perms)
        if dto.capabilities is not None:
            sets.append("capabilities = :capabilities::text[]")
            params["capabilities"] = dto.capabilities
        if dto.hooks is not None:
            sets.append("hooks = :hooks::text[]")
            params["hooks"] = dto.hooks
        if dto.tags is not None:
            sets.append("tags = :tags::text[]")
            params["tags"] = dto.tags
        if dto.status is not None:
            sets.append("status = :status")
            params["status"] = dto.status.value if hasattr(dto.status, "value") else dto.status

        if not sets:
            return await self.get_plugin(plugin_id, db)

        sets.append("updated_at = :updated_at")
        params["updated_at"] = datetime.now(timezone.utc)

        stmt = text(f"""
            UPDATE plugins SET {', '.join(sets)}
            WHERE id = :id
            RETURNING id, name, version, author, description, license, status,
                      permissions, capabilities, hooks, tags, entry_point,
                      tenant_id, homepage, repository, icon, config_schema, code,
                      min_app_version, created_at, updated_at
        """)
        result = await db.execute(stmt, params)
        row = result.mappings().first()
        return dict(row) if row else None

    async def delete_plugin(self, plugin_id: uuid.UUID, db: AsyncSession):
        stmt = text("DELETE FROM plugins WHERE id = :id")
        result = await db.execute(stmt, {"id": plugin_id})
        return result.rowcount > 0

    async def get_permissions(self, plugin_id: uuid.UUID, db: AsyncSession):
        stmt = text("SELECT permissions FROM plugins WHERE id = :id")
        result = await db.execute(stmt, {"id": plugin_id})
        row = result.mappings().first()
        return row["permissions"] if row else None

    async def execute(self, plugin_id: uuid.UUID, dto, db: AsyncSession):
        plugin = await self.get_plugin(plugin_id, db)
        if not plugin:
            raise HTTPException(status_code=404, detail="Plugin not found")
        if plugin["status"] == "DISABLED":
            raise HTTPException(status_code=403, detail="Plugin is disabled")

        code = dto.code or plugin["code"]
        if not code:
            raise HTTPException(status_code=400, detail="No code to execute")

        execution_id = uuid.uuid4()
        permissions = plugin.get("permissions", {})
        sandbox_context = {
            "input": dto.input or {},
            "tenantId": plugin.get("tenant_id"),
            "pluginId": str(plugin_id),
            "permissions": permissions,
        }

        start = time.monotonic()
        status = "completed"
        output = None
        error_message = None
        timeout_ms = dto.timeout_ms or 5000

        try:
            restricted_globals = {
                "__builtins__": {
                    "abs": abs, "all": all, "any": any, "bool": bool,
                    "dict": dict, "enumerate": enumerate, "filter": filter,
                    "float": float, "int": int, "isinstance": isinstance,
                    "len": len, "list": list, "map": map, "max": max,
                    "min": min, "range": range, "round": round, "set": set,
                    "slice": slice, "sorted": sorted, "str": str, "sum": sum,
                    "tuple": tuple, "type": type, "zip": zip, "reversed": reversed,
                    "True": True, "False": False, "None": None,
                    "Exception": Exception, "ValueError": ValueError,
                    "TypeError": TypeError, "KeyError": KeyError,
                    "IndexError": IndexError, "AttributeError": AttributeError,
                    "RuntimeError": RuntimeError, "StopIteration": StopIteration,
                },
                "input": sandbox_context["input"],
                "context": sandbox_context,
            }
            local_scope = {}
            exec(code, restricted_globals, local_scope)
            result = local_scope.get("result", local_scope.get("run", lambda: None)())
            output = {"result": result} if result is not None else None
        except Exception as e:
            status = "failed"
            error_message = f"{type(e).__name__}: {str(e)}"

        duration = int((time.monotonic() - start) * 1000)

        exec_stmt = text("""
            INSERT INTO plugin_executions (id, plugin_id, status, duration_ms, output, error_message, created_at)
            VALUES (:id, :plugin_id, :status, :duration_ms, :output::jsonb, :error_message, :created_at)
            RETURNING id, plugin_id, status, duration_ms, output, error_message, created_at
        """)
        exec_result = await db.execute(exec_stmt, {
            "id": execution_id,
            "plugin_id": plugin_id,
            "status": status,
            "duration_ms": duration,
            "output": json.dumps(output) if output else None,
            "error_message": error_message,
            "created_at": datetime.now(timezone.utc),
        })
        row = exec_result.mappings().first()
        return dict(row) if row else None

    async def list_executions(self, plugin_id: uuid.UUID, page: int, page_size: int, db: AsyncSession):
        offset = (page - 1) * page_size
        count_stmt = text("SELECT COUNT(*) FROM plugin_executions WHERE plugin_id = :plugin_id")
        count_result = await db.execute(count_stmt, {"plugin_id": plugin_id})
        total = count_result.scalar_one()

        stmt = text("""
            SELECT id, plugin_id, status, duration_ms, output, error_message, created_at
            FROM plugin_executions
            WHERE plugin_id = :plugin_id
            ORDER BY created_at DESC
            LIMIT :limit OFFSET :offset
        """)
        result = await db.execute(stmt, {"plugin_id": plugin_id, "limit": page_size, "offset": offset})
        rows = result.mappings().all()
        return [dict(r) for r in rows], total


plugin_service = PluginService()
